[TensorExpr] Fuser: only require input shapes to be known (output shapes can be inferred).

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43603 Benchmarks: temporarily disable profiling-te configuration.
* #43173 [TensorExpr] Start using typecheck in the fuser.
* #43235 [JIT] Subgraph utils: add an optional vmap argument to the API to allow retrieving value mappings.
* #43365 [TensorExpr] Fuser: disallow aten::slice nodes.
* **#43171 [TensorExpr] Fuser: only require input shapes to be known (output shapes can be inferred).**
* #43170 [TensorExpr] Fuser: move canHandle and some other auxiliary functions into TensorExprFuser class.



Differential Revision: [D23178228](https://our.internmc.facebook.com/intern/diff/D23178228)

Commit:
[TensorExpr] Fuser: only require input shapes to be known (output shapes can be inferred).

[ghstack-poisoned]

eellison reviewed (commented):
Why are we failing instead of bailing now?

ZolotukhinM commented:
> Why are we failing instead of bailing now?

Because the logic is that we are merging new nodes into an existing fusion group - all the nodes in the fusion group have to be supported, that's an invariant (we start from a fusible node and only add other fusible nodes there). This check should not be needed at all, but since we've had it before I replaced it with an assert.

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -309,11 +309,6 @@ class TensorExprFuser {
 
   bool allShapesAreKnown(Node* node) {
     // TODO: Relax the checks to support dynamic shapes
-    for (Value* output : node->outputs()) {
```
Just because you know the input shapes doesnt mean you know the output shapes:
```
def index(i: int):
    x = torch.tensor(1, 2, 3, 4)
    out = x[i]
```

ZolotukhinM replied:
It does for the subset of ops NNC supports though, at least currently. When we add more complicated ops, we would need to adjust the fuser logic as well, but as it stands now, I think we always can infer output shapes when we know input shapes.

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -353,17 +348,12 @@ class TensorExprFuser {
     // Symbolic checks
     REQ(canHandle(producer));
-    REQ((canHandle(consumer) || consumer->kind() == getTensorExprSymbol()));
+    AT_ASSERT(canHandle(consumer) || consumer->kind() == getTensorExprSymbol());
```
nit: `AT_ASSERT` -> `TORCH_INTERNAL_ASSERT`

eellison reviewed (commented):
Agreed the check in canMerge is redundant, I don't know about removing the output shapes check

Commit:
Update on "[TensorExpr] Fuser: only require input shapes to be known (output shapes can be inferred)."




Differential Revision: [D23178228](https://our.internmc.facebook.com/intern/diff/D23178228)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Fuser: only require input shapes to be known (output shapes can be inferred)."




Differential Revision: [D23178228](https://our.internmc.facebook.com/intern/diff/D23178228)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Fuser: only require input shapes to be known (output shapes can be inferred)."




Differential Revision: [D23178228](https://our.internmc.facebook.com/intern/diff/D23178228)

[ghstack-poisoned]
